src/main/kotlin/com/sourcegraph/cody/sidebar/WebThemeController.kt

```diff
@@ -0,0 +1,42 @@
+package com.sourcegraph.cody.sidebar
+
+import com.intellij.openapi.application.invokeLater
+import com.sourcegraph.config.ThemeUtil
+import java.awt.Color
+import javax.swing.UIManager
+
+class WebThemeController {
+  private var themeChangeListener: ((WebTheme) -> Unit)? = null
+
+  init {
+    UIManager.addPropertyChangeListener { event ->
+      if (event.propertyName == "lookAndFeel") {
+        invokeLater { themeChangeListener?.invoke(getTheme()) }
+      }
+    }
+  }
+
+  fun setThemeChangeListener(listener: (WebTheme) -> Unit) {
+    themeChangeListener = listener
+  }
+
+  fun getTheme(): WebTheme {
+    val themeVariables = mutableMapOf<String, String>()
+    val defaults = UIManager.getDefaults()
+    for (entry in defaults) {
+      val value = entry.value
+      if (value is Color) {
+        themeVariables[toCSSVariableName("${entry.key}")] = toCSSColor(value)
+      }
+    }
+    return WebTheme(ThemeUtil.isDarkTheme(), themeVariables)
+  }
+
+  private fun toCSSColor(value: Color) =
+      "rgb(${value.red} ${value.green} ${value.blue} / ${value.alpha / 255})"
+
+  private fun toCSSVariableName(key: String) =
+      "--jetbrains-${key.replace(Regex("[^-_a-zA-Z0-9]"), "-")}"
```

Comment:
For what cases exactly we need that regex?

Reply:
We need the generated identifiers to be valid CSS names, and some of the names contain periods. The regex is basically an allowlist (as you see) of valid CSS identifier characters, and anything outside that is replaced with -. Yes, this could collide in theory but those are inconvenient, not show stoppers--you will just get the last one.